src/ORM/ResultSetFactory.php

```diff
@@ -242,6 +249,45 @@ protected function groupResult(array $row, array $data): EntityInterface|array
         return $results;
     }
 
+    /**
+     * Cached DtoMapper instance
+     *
+     * @var \Cake\ORM\DtoMapper|null
+     */
+    protected ?DtoMapper $dtoMapper = null;
+
+    /**
+     * Hydrate a row into a DTO.
+     *
+     * @param array $row Nested array data
+     * @param class-string $dtoClass DTO class name
+     * @return object
+     */
+    protected function hydrateDto(array $row, string $dtoClass): object
+    {
+        // Check for array style static factory method
+        if (method_exists($dtoClass, 'createFromArray')) {
+            return $dtoClass::createFromArray($row, true);
+        }
+
+        // Use DtoMapper for plain readonly DTOs with named constructor params
+        return $this->getDtoMapper()->map($row, $dtoClass);
```

Comment:
This is a nice way to give developers an extension point to optimize things if they need it. Should there be an interface for `createFromArray` or is a method check good enough?